File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int ans = 0;

        if (k <= 1) {
            return ans;
        }

        for (int i = 0; i < nums.size(); i++) {
            int curr = 1;
            for (int j = i; j >= 0; j--) {
                curr *= nums[j];

                if (curr >= k) {
                    break;
                }

                ans++;
            }
        }

        return ans;
    }
};

File: 0713-subarray-product-less-than-k/0713-subarray-product-less-than-k.go
func numSubarrayProductLessThanK(nums []int, k int) int {
    l, r := 0, 0
    curr := 1
    ans := 0

    for r < len(nums) {
        if k <= 1 {
            return 0
        }

        curr *= nums[r]

        // maintain window
        for curr >= k {
            curr /= nums[l]
            l++
        }

        // 1 for subarray contains only r
        // r - l for subarray contains r and the rest nums within window
        ans += (r - l) + 1
        r++
    }

    return ans
}